### PR TITLE
fix(cfn-include): handle numbers expressed as strings in templates

### LINKED ISSUE #9524: Cfn-Include should accept strings where numbers are expected
CfnInclude throws type errors on templates that cloudformation accepts.

Any template that uses a string value where cfn include expects a number throws an error: eg

`CorsMaxAge: "10"` is valid cloudformation, but rejected by cfn include.

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by comcalvi**
Closes #9524

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/core/lib/runtime.ts [hunk 1]
```diff
@@ -217,7 +217,7 @@ export function validateString(x: any): ValidationResult {
 }
 
 export function validateNumber(x: any): ValidationResult {
-  if (canInspect(x) && typeof x !== 'number') {
+  if (canInspect(x) && (typeof x !== 'number' && typeof x !== 'string')) {
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/runtime.ts:220]: This is definitely not enough.

A string that doesn't parse to a decimal number (like `"abc") should not be correctly validated here.

File: packages/@aws-cdk/core/lib/runtime.ts [hunk 2]
```diff
@@ -218,8 +218,11 @@ export function validateString(x: any): ValidationResult {
 
 export function validateNumber(x: any): ValidationResult {
   if (canInspect(x) && typeof x !== 'number') {
-    return new ValidationResult(`${JSON.stringify(x)} should be a number`);
+    if (typeof x === 'string' && isNaN(parseInt(x))) {
```
Comment (rix0rrr) [inline packages/@aws-cdk/core/lib/runtime.ts:221]: Why must it be an int? Is `"3.141592654"` not allowed? Why does `"10"` get preferential treatment?

File: packages/@aws-cdk/cloudformation-include/test/test-templates/parsing-as-numbers.json [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+      "Properties": {
+        "CorsConfiguration": {
+          "CorsRules": [
+            {
+              "AllowedMethods": [
+                "GET"
+              ],
+              "AllowedOrigins": [
+                "*"
+              ],
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/test-templates/parsing-as-numbers.json:14]:
```suggestion
              "AllowedMethods": ["GET"],
              "AllowedOrigins": ["*"],
```

File: packages/@aws-cdk/cloudformation-diff/lib/diff/util.ts [hunk 4]
```diff
@@ -11,6 +11,7 @@
  */
 export function deepEqual(lvalue: any, rvalue: any): boolean {
   if (lvalue === rvalue) { return true; }
+  if (parseFloat(lvalue) === parseFloat(rvalue)) { return true; }
```
Comment (rix0rrr) [inline packages/@aws-cdk/cloudformation-diff/lib/diff/util.ts:14]: I had to double check some edge cases here, but it seems this is fine to do.

For the benefit of future scholars, can you add a comment line explaining why we're doing this?

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 5]
```diff
@@ -67,8 +67,16 @@ export class FromCloudFormation {
...
+    // return a number, if the input can be parsed as one
+    let parsedValue;
+    if (typeof value === 'string') {
+      parsedValue = parseFloat(value);
+      if (!isNaN(parsedValue)) {
+        return parsedValue;
+      }
+    }
+
+    // otherwise, return the input
```
Comment (rix0rrr) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:79]: ...and let the validator handle the fact that it's not the type we want.

That's a pretty important detail to understanding the behavior of the parser here. At first glance this code would be wrong, but it becomes correct in combination with some other code somewhere else.

**Pull request merged**